### PR TITLE
feat: Better Visual integration for music files 

### LINKED ISSUE #5228: Better Visual integration for music files
UI change
image
- song cover there

# Player
image
### pause/play button 
image
- hover effect as normal secondary button

### Time left
- time remaining of the mp4 

### waveform 
- showing progress of the song
- option to select part of the song, so its clickable 
> we can try in a first step to add progress bar and then try to add this waveform.

alternative progress bar: 
image

###  Mute option
image
- off/on audio - the icon changes to the speaker without the line 

# Card
image
- just little icon in the top left corner indicating that its a music nft

# Figma
- updated in gallery item handoff
- https://www.figma.com/file/gHSzqGPeQ9qK4L0qIDcTjY/gallery-item-handoff?node-id=1270%3A5615&t=5m2qKbEFYUwclsID-1
image

### LINKED ISSUE #6108: Audio Resource follow-up
hello, good job! The audio is already working. If it could be:
1. put there an auto-stop when the second audio starts
2. cover image could be in the background of the player 

https://github.com/kodadot/nft-gallery/assets/81104034/9e0e3851-37e8-4667-94db-25dc94090203

_Originally posted by @jkdizajn in https://github.com/kodadot/nft-gallery/issues/6031#issuecomment-1563404515_

### LINKED ISSUE #6237: Music nfts card visual
image
- lets match the style to the cards around
- player to the middle of the "img" preview 
https://canary.kodadot.xyz/ksm/gallery/13943908-cc89eac7f82ae14164-BIRBTAPES

### LINKED ISSUE #7272: Items list minor UX issues
### What happened?

1. When hovering over an NFT that is an audio file and clicking the audio controls the user is redirected to the item. The controls should only be shown if it can play the file. A good alternative would be to start playing the audio file after hovering for a longer time e.g. 3 seconds.
2. In case of audio files the item height is reduced, which looks bad

### Please reproduce in steps

Go to https://kodadot.xyz/rmrk/explore/items?listed=true

### Expected Behavior

Audio file should play when interacting with it's controls

### What browsers are you seeing the problem on?

Desktop Chrome

### At which address did you encounter bug?

_No response_

### Are you logged in?

None

### Which wallet you are using?

_No response_

### At which chain did you encounter bug?

_No response_

### Screenshots

Screenshot from 2023-09-14 16-56-55

### Relevant log output

_No response_

### Payment link for reward

https://kodadot.xyz/ksm/transfer/?target=%3CMy_Kusama_Address_check_https://github.com/kodadot/nft-gallery/blob/main/CONTRIBUTING.md#creating-your-ksm-address%3E

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### LINKED ISSUE #7292: Music files playing over each other
### What happened?

Go to: 
https://canary.kodadot.xyz/ksm/gallery/15297998-900d19dc7d3c444e4c-PLANETPO-DEEPOLOGIC_PLANET_PO_EP-00000004
play first song -> go to second one, play it.

https://github.com/kodadot/nft-gallery/assets/36627808/e9abdbda-b31c-42a5-89b9-3c711ca314dc

### Please reproduce in steps

a

### Expected Behavior

a

### What browsers are you seeing the problem on?

_No response_

### At which address did you encounter bug?

_No response_

### Are you logged in?

None

### Which wallet you are using?

_No response_

### At which chain did you encounter bug?

_No response_

### Screenshots

a

### Relevant log output

_No response_

### Payment link for reward

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by hassnian**
**Thank you for your contribution** to the [KodaDot - One Stop Shop for Polkadot NFTs](https://kodadot.xyz).

👇 __ Let's make a quick check before the contribution.

## PR Type

- [x] Bugfix
- [x] Feature
- [ ] Refactoring

## Context

test 

https://deploy-preview-7275--koda-canary.netlify.app/ksm/gallery/15297998-900d19dc7d3c444e4c-PLANETPO-DEEPOLOGIC_PLANET_PO_EP-00000004

https://deploy-preview-7275--koda-canary.netlify.app/bsx/collection/2338064950

- [x] Closes #5228
- [x] Closes #7272
- [x] Closes #7292 
- [x] Closes #6237
- [x] Closes #6108 

#### Did your issue had any of the "$" label on it?

- [x] Fill up your DOT address: [Payout](https://kodadot.xyz/dot/transfer?target=13QUj3pZyFNfYj4AM336hRdyLQbevj5H3sR4PKmLEXLdwZhh)

#### Community participation

- [x] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

## Screenshot 📸

- [x] My fix has changed **something** on UI; 

gallery item nft with audio resources
CleanShot 2023-09-17 at 14 56 32

gallery item nft audio 
CleanShot 2023-09-17 at 14 57 51

gallery item more about this collection 
CleanShot 2023-09-17 at 14 59 03@2x

collection page
CleanShot 2023-09-17 at 14 55 48@2x

### <samp>🤖 Generated by Copilot at 70a2946</samp>

Added a custom audio player component `NeoAudioPlayer` to improve the UI and UX of playing audio NFTs. Refactored `AudioMedia.vue` to use the new component and removed unused styles. The new component uses the `useMediaControls` hook and emits a `change` event for seeking.

### <samp>🤖 Generated by Copilot at 70a2946</samp>

> _`NeoAudioPlayer`_
> _A custom component for sound_
> _Spring of user joy_

### CODE REVIEW

**Comment by roiLeo** [comment]
eg:
- https://deploy-preview-7275--koda-canary.netlify.app/bsx/collection/2338064950

**Comment by hassnian** [comment]
@exezbcz visual check pls

File: libs/ui/src/components/NeoAudioPlayer/NeoAudioPlayer.vue [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+})
+
+const togglePlay = async () => {
+  if (playing.value) {
+    await pause()
+  } else {
+    await play()
+  }
+}
+
```
Comment (hassnian) [inline libs/ui/src/components/NeoAudioPlayer/NeoAudioPlayer.vue:92]: https://developer.chrome.com/blog/play-request-was-interrupted/

File: libs/ui/src/components/NeoAudioPlayer/NeoAudioPlayer.vue [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+const flushPreviouseActions = async () => {
+  await actionStack.value.map((action) => action.reject())
+  actionStack.value = []
+}
+
+const pushToActionStack = ({ promise, reject }) => {
```
Comment (hassnian) [inline libs/ui/src/components/NeoAudioPlayer/NeoAudioPlayer.vue:98]: fix for https://developer.chrome.com/blog/play-request-was-interrupted/

File: libs/ui/src/components/NeoAudioPlayer/NeoAudioPlayer.vue [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+const togglePlay = async () => {
+  if (playing.value) {
+    await pause()
+  } else {
+    await play()
+  }
+}
+
+const flushPreviouseActions = async () => {
```
Comment (hassnian) [inline libs/ui/src/components/NeoAudioPlayer/NeoAudioPlayer.vue:93]: fix for:
card hover in ->play
card hover out -> still play promise is being called need to flush and then stop 
The play() request was interrupted by a new load request. https://goo.gl/LdLk22

**Comment by exezbcz** [comment]
> @exezbcz visual check pls

I will check today, thanks!

**Comment by exezbcz** [comment]
can we please enable skipping/clicking on the progress bar. 
image


btw have you tried adding the waveform or we will leave it for followup?


one more thing, I think the use of this icon is now switched. Or at least it appears to me like that. 
image
- when I see muted icon, for me the current music is muted - or at least that how it should be. Wdyt? 
image

### hover 
- please add button hovers 
- use secondary button as reference
image
image

**Comment by prury** [comment]
missing icon on cards: 
image

**Comment by hassnian** [comment]
> missing icon on cards: 
> 
> image
> 
> 

here I wasn't sure , right now the icon is shown for nfts that **are a audio file** first 

not for nfts that have audio inside the resources if that's is the case should we also show the icon and also play maybe the first resources ?

I guess this is what you mean with missing icon

**Comment by hassnian** [comment]
> @hassnian is it hard to fix #7292 as well? if yes, leave if it for another PR
> 
> 


sure I'll fix that

> 
> when you enter inside gallery item for the first time, the mute button state is active but you can hear the song anyway:
> 
> image
> 
> 
> 
> edit: its function is actually inverted
> 
> 

yep mb I thought It was like the play pause button where the icon that you see is the action but it guess is not lul , will fix

**Comment by hassnian** [comment]
> can we please enable skipping/clicking on the progress bar. image
> 

hmm is not working for you ? it's enabled 



> btw have you tried adding the waveform or we will leave it for followup?
> 

I saw this comment 
CleanShot 2023-09-19 at 11 27 42@2x
we can leave it for a follow up

but i did play around a little bit the package the we currently use , didn't allow me to get the ui exactly as you would want , also found [this](https://codeacme17.github.io/1llest-waveform-vue/) other package that looks also good , its vue3 though,  we  might want to wait for the nuxt3 migration or just look for another package or some other alternative   

> one more thing, I think the use of this icon is now switched. Or at least it appears to me like that. image
> 
> * when I see muted icon, for me the current music is muted - or at least that how it should be. Wdyt?
> 
> image

yep that makes more sense , fixed 

> ### hover
> * please add button hovers
> * use secondary button as reference
> 
> image image

fixed

**Comment by hassnian** [comment]
@prury pls try again all the comment are fixed including #7292 

also check this https://github.com/kodadot/nft-gallery/pull/7275#issuecomment-1723716253 

 @roiLeo what do you think ?

**Comment by roiLeo** [comment]
> @roiLeo what do you think ?

> I saw this comment CleanShot 2023-09-19 at 11 27 42@2x we can leave it for a follow up

As someone who buys music a lot, I find that waveform is mandatory. 
Depends on how much times it would take to integrate but we can do a follow up.

> but i did play around a little bit the package the we currently use , didn't allow me to get the ui exactly as you would want , also found [this](https://codeacme17.github.io/1llest-waveform-vue/) other package that looks also good , its vue3 though, we might want to wait for the nuxt3 migration or just look for another package or some other alternative

hmmmm... This package is buggy for me, no need to have vue3 only package we can try to use something like wavesurfer.js.


Just tested on deeplogic examples and it's really nice. Good work!
Hopefully nothing else has changed or broken since you've touched many sensitive parts (MediaItem, NeoButton, NeoNftCard & nftList query)

**Comment by hassnian** [comment]
> wavesurfer.js. 

yep also I had this one in mind , was able to get the exact ui modifying a couple of the props 

I meant if you could check this comment https://github.com/kodadot/nft-gallery/pull/7275#issuecomment-1723716253

it's regarding nfts that are not a sound but have a sound resources

**Comment by roiLeo** [comment]
> here I wasn't sure , right now the icon is shown for nfts that **are a audio file** first , indicating that you can hover and play

Alright!

> not for nfts that have audio inside the resources if that's is the case should we also show the icon and also play maybe the first resources ?

I don't mind at first if it's not displayed, I'd like it to be based on the type of the metadata but we need to fix it on indexer first  (kodadot/rubick#104)

**Comment by hassnian** [comment]
> > here I wasn't sure , right now the icon is shown for nfts that **are a audio file** first , indicating that you can hover and play
> 
> Alright!
> 
> > not for nfts that have audio inside the resources if that's is the case should we also show the icon and also play maybe the first resources ?
> 
> I don't mind at first if it's not displayed, I'd like it to be based on the type of the metadata but we need to fix it on indexer first ([kodadot/rubick#104](https://github.com/kodadot/rubick/issues/104))

yep otherwise this is how it's done atm , bcz I think the ghraphql query returns "" 

CleanShot 2023-09-19 at 16 45 58@2x

and you can see how many requests it would have to make for each nft -> resources

**Comment by prury** [comment]
> > missing icon on cards:
> > image
> 
> here I wasn't sure , right now the icon is shown for nfts that **are a audio file** first , indicating that you can hover and play
> 
> not for nfts that have audio inside the resources if that's is the case should we also show the icon and also play maybe the first resources ?
> 
> I guess this is what you mean with missing icon

Sorry for the delay, seems that roiLeo already got ya.
will check PR again now

**Comment by prury** [comment]
play on hover ✔️ 
gallery music items ✔️ 
carousel items ✔️ 
buttons seems OK ✔️ 
user profile show it fine ✔️ 
explorer shows it fine ✔️ 
card transitions ✔️ 
play stop button ✔️ 
mute button ✔️ 
music pauses if you play another one ✔️ 

another collection for testing: https://deploy-preview-7275--koda-canary.netlify.app/ksm/collection/3629a3ff458d113e25-MUSIC?page=2

unrelated: music files takes a looong time to load and even refreshing doesn't do the job sometimes

**Comment by daiagi** [comment]
still no icon on kusama
image

**Comment by hassnian** [comment]
@daiagi 
1 - this is to be expected at this moment the nft resource type is an image and not a audio file those who have an audio file as the main resource will have that icon atm

ref https://github.com/kodadot/nft-gallery/pull/7275#issuecomment-1730829955 https://github.com/kodadot/nft-gallery/pull/7275#issuecomment-1725345866


CleanShot 2023-09-23 at 11 31 09@2x

**Comment by exezbcz** [comment]
@daiagi yes, then we can show the loading icon after user clicks the play button? 

Two scenarios: 

- music still loading, user did not click play - play button displayed, when its clicked, it will become the loading wheel 

- music ready, user did not click play - play button displayed, when clicked, music is playing.

File: libs/ui/src/components/NeoAudioPlayer/NeoAudioPlayer.vue [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+}
+
+const goToEnd = () => {
+  if (!canStartPlaying.value) {
+    return
+  }
+  play(duration.value)
+}
+
+useEventListener(audio, 'canplaythrough', () => {
```
Comment (daiagi) [inline libs/ui/src/components/NeoAudioPlayer/NeoAudioPlayer.vue:175]: How about change this event to 'canplay'?

That should give the feeling of quick loading and it willing be online with exez suggestion


https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/canplay_event

**Comment by daiagi** [comment]
>music still loading, user did not click play - play button displayed, when its clicked, it will become the loading wheel


It will be wierd to have loading spinner instead of pause after clicking play

Can we consider separating the loading indicator away from the play button?

**Comment by vikiival** [comment]
> Hopefully nothing else has changed or broken since you've touched many sensitive parts (MediaItem, NeoButton, NeoNftCard & nftList query)

OOF. Like honestly the target of issue was to make Gallery Item usable, Everything else can be done as a follow-up

File: libs/ui/src/components/NeoAudioPlayer/NeoAudioPlayer.vue [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+  if (playing.value) {
+    await pause()
+  } else {
+    await play()
+  }
+}
+
+const flushPreviouseActions = async () => {
+  try {
+    await actionStack.value.map((action) => action.reject())
```
Comment (daiagi) [inline libs/ui/src/components/NeoAudioPlayer/NeoAudioPlayer.vue:102]: this here is not a promise, can't await on it

image

**Review by daiagi — APPROVED** [review]
nice work
let's deal with the loading indicator in a follow up

File: libs/ui/src/components/NeoAudioPlayer/NeoAudioPlayer.vue (hunk 5 above)
Comment (hassnian) [reply libs/ui/src/components/NeoAudioPlayer/NeoAudioPlayer.vue:102]: missed the Promise.all mb

**Comment by yangwao** [comment]
cc @exezbcz

**Comment by exezbcz** [comment]
letsgo, works for me, I will create followups

**Comment by yangwao** [comment]
pay 70 usd for extra bravery to close down 5 issues 💪

@hassnian let's resolve conflicts and it's good to go for the merge

**Pull request merged**